### PR TITLE
Tests and fixes for seeking and skipping in audio streams

### PR DESCRIPTION
**Pull request by uklotzde**
I added the following test cases for SoundSources:
- Seek and read samples (repeatedly and alternating)
- Seek to stream boundaries followed by verified reading from the middle of the stream

Apparently various implementations were broken! All fixed with one exception: Seeking near the end of an MP3 stream fails with SoundSourceMP3 (libmad). I temporarily excluded/disabled the failing test and marked it with a TODO. I have no idea what's going wrong here.

The fix for SoundSourceOpus is the biggest one, because opusfile neither implements skipping nor provides sample accurate seeking and decoding. By prefetching sample frames while seeking the accuracy can be improved substantially. All tunable parameters that affect the sample accuracy are now defined in SoundSourceOpus and are available as constants for verifying the test results.

### CODE REVIEW

**Comment by uklotzde** [comment]
Some additional commits (trial and error) were necessary to fix the build and tests for all supported platforms.

File: src/util/audiosignal.cpp [hunk 1]
```diff
@@ -4,6 +4,31 @@
...
 namespace mixxx {
 
+// Definitions of static class constants are required by
+// LLVM CLang, but cause "multiply defined symbols" errors
+// in Visual Studio 2013. GCC is happy either with or
+// without those definitions.
+// TODO(XXX): Replace 'static const' with 'static constexpr'
+// in the corresponding .h file after upgrading to Visual
+// Studio 2015 and remove all definitions from this .cpp file.
+#if !defined(_MSC_VER) || (_MSC_VER >= 1900)
```
Comment (rryan) [inline src/util/audiosignal.cpp]: "not Visual Studio OR Visual Studio 2015 and newer"

Is that what you intended? Did you mean "only Visual Studio earlier than 2015"? 

```
#if defined(_MSC_VER) && (_MSC_VER < 1900)
```

File: src/util/audiosignal.cpp [hunk 2]
```diff
@@ -4,6 +4,31 @@
 
 namespace mixxx {
 
+// Definitions of static class constants are required by
```
Comment (rryan) [inline src/util/audiosignal.cpp]: are not required?

File: src/util/audiosignal.cpp [hunk 3]
```diff
@@ -4,6 +4,31 @@
...
+// LLVM CLang, but cause "multiply defined symbols" errors
+// in Visual Studio 2013. GCC is happy either with or
```
Comment (rryan) [inline src/util/audiosignal.cpp]: GCC/Clang?

Comment (rryan) [inline src/util/audiosignal.cpp]: I'm confused -- how does this build at head (since Travis/Appveyor uses all of GCC, LLVM, and MSVC2013)?

**Review by rryan — COMMENTED** [review]
I skimmed the decoder changes and they look good -- since there is high test coverage I didn't do a thorough review. Thanks!

File: src/util/audiosignal.cpp (hunk 1 above)
Comment (uklotzde) [inline src/util/audiosignal.cpp]: GCC: With or without definitions, doesn't matter -> **optional***
LLVM: Definitions are **mandatory**
Visual Studio 2013: Definitions produce link errors -> **forbidden**

There is no intersection that works for all 3 compilers: I decided to add the definitions as a default with an exception for Visual Studio <= 2013. Starting with the Visual Studio 2015 the definitions will be included (as for GCC/LLVM) and we will see what happens.

Comment (rryan) [inline src/util/audiosignal.cpp]: Hm ok. But I'm still confused why this builds successfully at head?

**Comment by uklotzde** [comment]
!defined(_MSC_VER) -> GCC, LLVM
_MSC_VER >= 1900   -> Visual Studio 2015, 2017, ...

File: src/util/audiosignal.cpp (hunk 1 above)
Comment (rryan) [inline src/util/audiosignal.cpp]: What's confusing is that master builds fine with clang and the static constants defined inline in audiosignal.h and not in audiosignal.o, which contradicts what you are saying about them being mandatory?

**Comment by uklotzde** [comment]
I have reset the branch and excluded the confusing workaround. Now the build for OS X fails:

```
[LD] osx64_build/mixxx-test

Undefined symbols for architecture x86_64:

   "mixxx::AudioSignal::kChannelCountStereo", referenced from:

       mixxx::SoundSourceOpus::tryOpen(mixxx::AudioSourceConfig const&) in soundsourceopus.o

ld: symbol(s) not found for architecture x86_64

clang: error: linker command failed with exit code 1 (use -v to see invocation)

scons: *** [osx64_build/mixxx-test] Error 1

scons: building terminated because of errors.

```

**Comment by rryan** [comment]
> I have reset the branch and excluded the confusing workaround. Now the build for OS X fails:

Very weird, so it must be some change within this PR that causes clang to need this? The constants were already used within multiple .o's before this PR.

**Comment by rryan** [comment]
Sorry @uklotzde , I didn't mean to make you waste time on this when what you had was working. I was just curious what the reason was. We can merge what you had if you force push it back.

**Comment by timrae** [comment]
On ubuntu 16.10, gcc (Ubuntu 6.2.0-5ubuntu12) 6.2.0 20161005
`scons -j8 optimize=off faad=1` fails with linking errors:

```bash
lin64_build/sources/soundsourceopus.o: In function `mixxx::SoundSourceOpus::tryOpen(mixxx::AudioSourceConfig const&)': /home/tim/development/mixxx/src/sources/soundsourceopus.cpp:205: undefined reference to `mixxx::AudioSignal::kChannelCountStereo' collect2: error: ld returned 1 exit status scons: *** [lin64_build/mixxx] Error 1 scons: building terminated because of errors.```

`optimize=native` works as expected

**Comment by uklotzde** [comment]
This should fix the linking issues on all platform:

https://github.com/mixxxdj/mixxx/pull/1139

As we've learned the definitions in the .cpp file are actually required. And 
Visual Studio 2015 does not complain about duplicate symbols anymore.

**Pull request merged**